unimport/refactor.py

```diff
@@ -1,137 +1,113 @@
-from contextlib import contextmanager
-from lib2to3.fixer_base import BaseFix
-from lib2to3.fixer_util import (
-    BlankLine, FromImport, Leaf, Newline, syms, token)
-from lib2to3.refactor import RefactoringTool
-
-
-def traverse_imports(names):
-    """
-    Walks over all the names imported in a dotted_as_names node.
-    """
-    pending = [names]
-    while pending:
-        node = pending.pop()
-        if node.type in {token.NAME, token.STAR}:
-            yield node.value
-        elif node.type == syms.dotted_name:
-            yield "".join([ch.value for ch in node.children])
-        elif node.type in {syms.dotted_as_name, syms.import_as_name}:
-            yield node.children[2].value
-        elif node.type in {syms.dotted_as_names, syms.import_as_names}:
-            pending.extend(node.children[::-2])
-        else:
-            raise ValueError("unknown node type", node.type)
-
-
-class RefactorImports(BaseFix):
-
-    PATTERN = r"""
-        simple_stmt<
-            (
-                import_name< 'import' imp=any >
-                |
-                import_from< 'from' imp=(['.'*] any) 'import' ['('] items=any [')'] >
-            ) '\n'
-        >
-    """
-
-    def __init__(self):
-        self.unused_modules = []
-        super().__init__(None, None)  # options and logger
-
-    @contextmanager
-    def clean(self, unused_modules):
-        try:
-            self.unused_modules.clear()
-            self.unused_modules.extend(unused_modules)
-            yield
-        finally:
-            self.unused_modules.clear()
-
-    def transform(self, node, results):
-        imp = self.get_imp_if_equal_to_lineno(node.get_lineno())
-        if imp:
-            if node.children[0].type == syms.import_from:
-                return self.import_from(node, results, imp)
-            elif node.children[0].type == syms.import_name:
-                return self.import_name(node, results)
-
-    def import_from(self, node, results, imp):
-        if imp["star"] and imp["module"]:
-            if not imp["modules"]:
-                return BlankLine()
+import libcst as cst
+from libcst.metadata import MetadataWrapper, PositionProvider
+
+from unimport.scan import Scanner
+
+
+class RemoveUnusedImportTransformer(cst.CSTTransformer):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+
+    def __init__(self, source):
+        self.scanner = Scanner(source)
+        self.unused_imports = list(self.scanner.get_unused_imports())
+        self.imports = self.scanner.imports
```

Comment:
This shouldn't instantiate the scanner, it can take relevant data from `Session`.